urllib3/util/rfc6555.py

```diff
@@ -0,0 +1,241 @@
+    from time import monotonic
+except ImportError:
+    from time import time as monotonic
+
+_HAPPY_EYEBALLS_CACHE = {}
+_HAPPY_EYEBALLS_CACHE_TIME = 60 * 10  # 10 minutes according to RFC 6555
+_ASYNC_ERROR_NUMBERS = set([errno.EINPROGRESS,
+                            errno.EAGAIN,
+                            errno.EWOULDBLOCK])
+if hasattr(errno, "WSAEWOULDBLOCK"):  # Platform-specific: Windows
+    _ASYNC_ERROR_NUMBERS.add(errno.WSAEWOULDBLOCK)
+
+_IP_FAMILIES = set([socket.AF_INET])
+if hasattr(socket, "AF_INET6"):
+    _IP_FAMILIES.add(socket.AF_INET6)
+
+
+def happy_eyeballs_algorithm(address, timeout=socket._GLOBAL_DEFAULT_TIMEOUT,
+                             source_address=None, socket_options=None):
+    """ Implements the Happy Eyeballs protocol (RFC 6555) which allows
+    multiple sockets to attempt to connect from different families
+    for better connect times for dual-stack clients where server
+    IPv6 service is advertised but broken. """
+    result = None  # This is the actual connected socket eventually.
+    err = None
+    family = 0
+    cached = False
+
+    # We need to keep track of the time so we don't exceed timeout.
+    start_time = monotonic()
+    if timeout is None:
+        timeout_time = None
+    elif timeout is socket._GLOBAL_DEFAULT_TIMEOUT:
+        timeout_time = socket._GLOBAL_DEFAULT_TIMEOUT
+    else:
+        timeout_time = start_time + timeout
+
+    # Check the cache to see if our address is already there.
+    if address in _HAPPY_EYEBALLS_CACHE:
+        family, expires = _HAPPY_EYEBALLS_CACHE[address]
+
+        # If the cache entry is expired, don't use it.
+        if start_time > expires:
+            del _HAPPY_EYEBALLS_CACHE[address]
+            family = 0
+        else:
+            cached = True
+
+    host, port = address
+    socks = []
+
+    # Make sure we close the selector after we're done.
+    with DefaultSelector() as selector:
+        # Perform a DNS lookup for the address for IPv4 or IPv6 families.
+        if not family:
+            family = socket.AF_UNSPEC
+
+        dns_results = socket.getaddrinfo(host, port, family, socket.SOCK_STREAM)
+        dns_results_len = len(dns_results)
+
+        for i in range(dns_results_len):
+            af, socktype, proto, canonname, sa = dns_results[i]
+
+            # We only care about IPv4 and IPv6 addresses.
+            if af not in _IP_FAMILIES:
+                continue
+
+            sock = None
+            try:
+                sock = socket.socket(af, socktype, proto)
+                if timeout_time is socket._GLOBAL_DEFAULT_TIMEOUT:
+                    sock_timeout = sock.gettimeout()
+                    if sock_timeout is None:
+                        timeout_time = None
+                    else:
+                        timeout_time = start_time + sock_timeout
+
+                if socket_options:
+                    for opt in socket_options:
+                        sock.setsockopt(*opt)
```

Comment:
Let's try to continue to use the utility function we wrote for this in `util.connection`. It may need to be hoisted into this module, but that's ok.